java-checks/src/main/java/org/sonar/java/checks/CollectionIsEmptyCheck.java

```diff
@@ -54,38 +57,62 @@
     .names("size")
     .addWithoutParametersMatcher()
     .build();
-
-  private JavaFileScannerContext context;
+  private static final Tree.Kind[] TARGETED_BINARY_OPERATOR_TREES = {
+    Tree.Kind.EQUAL_TO,
+    Tree.Kind.NOT_EQUAL_TO,
+    Tree.Kind.LESS_THAN,
+    Tree.Kind.LESS_THAN_OR_EQUAL_TO,
+    Tree.Kind.GREATER_THAN,
+    Tree.Kind.GREATER_THAN_OR_EQUAL_TO
+  };
+  private static final Tree.Kind[] CLASS_TREES = {
+    Tree.Kind.CLASS,
+    Tree.Kind.ENUM,
+    Tree.Kind.INTERFACE,
+    Tree.Kind.RECORD,
+    Tree.Kind.ANNOTATION_TYPE
+  };
+  private static final Deque<Boolean> IS_COLLECTION_STACK = new LinkedList<>();
 
   @Override
-  public void scanFile(JavaFileScannerContext context) {
-    this.context = context;
+  public List<Tree.Kind> nodesToVisit() {
+    return ListUtils.concat(Arrays.asList(CLASS_TREES), Arrays.asList(TARGETED_BINARY_OPERATOR_TREES));
+  }
 
-    scan(context.getTree());
+  @Override
+  public void leaveFile(JavaFileScannerContext context) {
+    // in case of polutated state
+    IS_COLLECTION_STACK.clear();
   }
 
   @Override
-  public void visitClass(ClassTree tree) {
-    for (Tree member : tree.members()) {
-      if (!tree.symbol().type().isSubtypeOf(JAVA_UTIL_COLLECTION) || !member.is(Tree.Kind.METHOD)) {
-        scan(member);
-      }
+  public void visitNode(Tree tree) {
+    if (tree.is(CLASS_TREES)) {
+      boolean isCollection = ((ClassTree) tree).symbol().type().isSubtypeOf(JAVA_UTIL_COLLECTION);
+      IS_COLLECTION_STACK.push(isCollection);
+      return;
+    }
+    // Necessarily a BinaryExpressionTree - the rule only raise issues when size() comparison
+    // are "poorly" used from something which is NOT a Collection
+    if (Boolean.FALSE.equals(IS_COLLECTION_STACK.peek())) {
```

Comment:
Should we rewrite the test as a separate method and exit early?
```
if (isNodeInACollectionType()) {
  return;
}
```

Reply:
I think the test can be misled because we may be testing size() against 0 in an inner class.
```
public class MyCollectionType<T> implements Collection<T> {

  @Override
  public boolean isEmpty() {
    return false;
  }

  @NotNull
  @Override
  public Iterator<T> iterator() {
    return new MyIterator<>();
  }
  
  class MyIterator<T> implements Iterator<T> {

    @Override
    public boolean hasNext() {
      return MyCollectionType.this.size() == 0;
    }
  }
}
```
Should we consider this case?
```suggestion
    if (IS_COLLECTION_STACK.stream().allMatch(Boolean.FALSE::equals)) {
```

Reply:
damned. This question of inner class vs nested class is making the rule much more complex than expected.